docs/components/material.md

```diff
@@ -378,6 +451,137 @@ AFRAME.registerShader('custom', {
+
+## Supported Uniform Types
+
+The uniform types supported by A-Frame are summarized in the table below.
+Note that `time` can eliminate the need for a `tick()` handler in many cases.
+
+| A-Frame Type | THREE Type | GLSL Shader Type |
+|--------------|------------|------------------|
+| array | v3 | vec3 |
+| color | v3 | vec3 |
+| int   | i  | int  |
+| number | f | float |
+| map | t | map |
+| time | f | float (milliseconds) |
+| vec2 | v2 | vec2 |
+| vec3 | v3 | vec3 |
+| vec4 | v4 | vec4 |
+
+### Example — GLSL and Shaders
+
+For more customized visual effects, we can write GLSL shaders and apply them to A-Frame entities. 
+
+> NOTE: GLSL, the syntax used to write shaders, may seem a bit scary at first. For a gentle (and free!) introduction, we recommend [The Book of Shaders](http://thebookofshaders.com/).
+
+Here are the vertex and fragment shaders we'll use:
+
+```glsl
+// vertex.glsl
+
+varying vec2 vUv;
+
+void main() {
+  vUv = uv;
+  gl_Position = projectionMatrix * modelViewMatrix * vec4( position, 1.0 );
+}
+```
+
+```glsl
+// fragment.glsl
+
+varying vec2 vUv;
+uniform vec3 color;
+uniform float time1000;
+
+void main() {
+  // Use sin(time), which curves between 0 and 1 over time,
+  // to determine the mix of two colors:
+  //    (a) Dynamic color where 'R' and 'B' channels come
+  //        from a modulus of the UV coordinates.
+  //    (b) Base color.
+  // 
+  // The color itself is a vec4 containing RGBA values 0-1.
+  gl_FragColor = mix(
+    vec4(mod(vUv , 0.05) * 20.0, 1.0, 1.0),
+    vec4(color, 1.0),
+    sin(time1000 / 1000.0) // since A-Frame time is in milliseconds
+  );
+}
+```
+
+To use these vertex and fragment shaders, 
+after reading them into strings `vertexShader` and `fragmentShader`,
+we simply register our custom shader with A-Frame:
+
+```js
+// shader-grid-glitch.js
+
+AFRAME.registerShader('grid-glitch', {
+  schema: {
+    color: {type: 'color', is: 'uniform'},
+    time1000: {type: 'time', is: 'uniform'}
+  },
+
+  vertexShader: vertexShader,
+  fragmentShader: fragmentShader
+});
+```
+
+Finally, here is the HTML markup to put it all together:
+
+```html
+<!-- index.html -->
+
+<a-scene>
+  <a-sphere material="shader:grid-glitch; color: blue;"
+            radius="0.5"
+            position="0 1.5 -2">
+  </a-sphere>
+</a-scene>
+```
+
+* [Live demo](https://aframe-simpler-shader.glitch.me/)
+* [Remix this on Glitch](https://glitch.com/edit/#!/aframe-simpler-shader)
+
+## Register a Custom Component Using THREE Material
+
+For those cases where the `registerShader` API lacks needed functionality 
+(e.g., no `tick` handler, some missing uniform types), 
+we recommend creating a custom material by
+creating three.js materials (e.g., `RawShaderMaterial`, `ShaderMaterial`) in a
+component:
+
+```js
+AFRAME.registerComponent('custom-material', {
+  schema: {
+    // Add properties.
+  },
+
+  init: function () {
+    this.material = this.el.getOrCreateObject3D('mesh').material = new THREE.ShaderMaterial({
```

Comment:
This predates your PR, but i think `getOrCreateObject3D('mesh').material` is a bad pattern. It screws with components waiting for `object3dset`, for example. IMO this initializer should see if a mesh is available already, and if not wait for the `object3dset` method.

(this has the added benefit of making it compatible with custom models, which `registerShader` is not afaik)

Reply:
How does it mess with components waiting for object3dset?

Reply:
See: https://github.com/aframevr/aframe/issues/2464

Reply:
Point is, the framework should abstract all this away so you can say `shader:displacement` and let A-frame worry about when.  That is what `registerShader` allows, and this entire approach completely fails to do.

Reply:
> if you use a traverse(), which registerShader cannot do
Shouldn't the `material` component be the one doing that?

Reply:
![whynotboth](https://user-images.githubusercontent.com/1848368/31422749-af0b11b6-ae04-11e7-8955-87b9d200b6d5.jpg)

Reply:
My thought was a simple `recursive:true` (or we can still call it traverse).  I am guessing we can detect the aframe entity object3D structure and refuse to cascade, but I haven't looked.

Reply:
My thought was a simple `recursive:true` (or we can still call it traverse).  I am guessing we can detect the aframe entity object3D structure and refuse to cascade, but I haven't looked.

Reply:
You can always make your own replacement for the material component, which obviously you do by default, but for the simple example cases that seems like overkill.

What I am not sure about is whether the model object trees have consistent DOM-ish naming, for better declarative mapping etc.  I think I saw discussion on that in another issue/PR

Reply:
I've found that when I need a custom material, I just use a raw component. I think @donmccurdy does the same? It's simpler than hooking into `material`. Should we recommend that?

Reply:
The idea is that if you only need to use your own (vertex and fragment) shader, that is most easily done with `registerShader` and then `material="shader:my-shader"`.  The examples that were given previously are actually easier to do that way (which is hopefully illustrated).

Other examples where you only need shader and not custom material, besides the ones listed, include the typical top/bottom and left/right stereo shaders we used to recommend separate geometry and/or components for.

If `registerShader` doesn't work for various reasons, then it does recommend creating a custom component (with this example). 

I did ask for an example that actually requires custom component, but so far no one has suggested one.

Reply:
Nevermind, I saw this is already done as a raw component. I thought it was trying registerShader.